### ISSUE TITLE
Warn logging with "Failed to mark a promise as success because it is done already: DefaultChannelPromise" after upgrading to 4.0.17.Final

### ISSUE THREAD

**Issue by twilek** [netty/netty#2302]
This may not be a bug, but I have a very hard time figuring out what is wrong. After upgrading from 4.0.15.Final to 4.0.17.Final I now get quite many warn messages in my log with the message "Failed to mark a promise as success because it is done already: DefaultChannelPromise" in my production environment. The server seems to work fine both before and after upgrade and I do not get any stack where this error is triggerd and can not really see what is wrong and how to get rid of this warning. Any suggestions?

**Comment by chemist777** [comment]
In latest Netty 5 i had similar exceptions.

Cause was this code fragment:
```
if (frame instanceof CloseWebSocketFrame) {
    ctx.channel().write(frame.retain()).addListener(ChannelFutureListener.CLOSE);
    return;
}
```
After this changes (flush) exceptions disappeared:
```
if (frame instanceof CloseWebSocketFrame) {
    ctx.channel().writeAndFlush(frame.retain()).addListener(ChannelFutureListener.CLOSE);
    return;
}
```

**Comment by clienthax** [comment]
Thats it, just keep bumping that milestone.

**Comment by trustin** [comment]
![bad taste to bad taste](https://cloud.githubusercontent.com/assets/173918/9402078/b539fe0c-4813-11e5-8e2e-9be76ef65009.png)

**Comment by cpw** [comment]
*milestone bump* Nice..

**Comment by trustin** [comment]
TL;DR - Closing, because it's a user error.

According to the original report from @twilek, marking the promise as success has failed because the promise has been failed already due to `IllegalReferenceCountException`. As the exception signifies, it means a user (or some component of Netty) attempted to decrease (or increase) the reference count of a buffer in an unexpected way.

What @lexand shared in his comment is one of the common user mistakes. He (or she) created a buffer as a static field and reused it over and over. When a buffer is written out to a socket, Netty releases the buffer (i.e. decreases its reference count.) As @lexand fixes his/her problem, you need to make such a buffer unreleasable or you have to call `retain()` on each write.

The reproducer that @maekitalo posted also has the same problem with what @lexand experienced. He or she created a `DefaultFullHttpRequest` and reused it without retaining it. Why `DefaultFullHttpRequest`? It's because a `DefaultFullHttpRequest` allocates its own content buffer unless you specified the content. If the content is empty, you could specify `Unpooled.EMPTY_BUFFER` explicitly. It's unreleasable.

I'm not sure if the problem @chemist777 posted is related with the original report. At least our WebSocket support code seems to be fixed now.

Other complaints about our milestone bumping..? Without more useful information, they are just begging to get marked as a 'user error'.
